File: python/day4b.py
#!/usr/bin/env python3

import re

with open("../input/day4.txt") as f:
    data = f.read()

def valid_byr(s):
    try:
        if len(s) != 4:
            return
        i = int(s)
        if i < 1920 or i > 2002:
            return False
        return True
    except:
        return False

def valid_iyr(s):
    try:
        if len(s) != 4:
            return
        i = int(s)
        if i < 2010 or i > 2020:
            return False
        return True
    except:
        return False

def valid_eyr(s):
    try:
        if len(s) != 4:
            return
        i = int(s)
        if i < 2020 or i > 2030:
            return False
        return True
    except:
        return False

def valid_hgt(s):
    try:
        if s.endswith('cm'):
            i = int(s[:-2])
            if i < 150 or i > 193:
                return False
            return True
        if s.endswith('in'):
            i = int(s[:-2])
            if i < 59 or i > 76:
                return False
            return True
        return False
    except:
        return False

re_hcl = re.compile('^#[0-9a-f]{6}$')
def valid_hcl(s):
    return bool(re_hcl.match(s))

def valid_ecl(s):
    return s in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']

re_pid = re.compile('^[0-9]{9}$')
def valid_pid(s):
    return bool(re_pid.match(s))

required_fields = {
    'byr': valid_byr,
    'iyr': valid_iyr,
    'eyr': valid_eyr,
    'hgt': valid_hgt,
    'hcl': valid_hcl,
    'ecl': valid_ecl,
    'pid': valid_pid,
}

count = 0
found_fields = {}

def finish_fields():
    global count
    global found_fields

    if 'cid' in found_fields:
        del found_fields['cid']
    if len(found_fields) == len(required_fields):
        valid = True
        for field, is_valid in required_fields.items():
            if not is_valid(found_fields[field]):
                valid = False
                break
        if valid:
            count += 1
    found_fields = {}

for entry in data.splitlines():
    entry = entry.strip()
    if not entry:
        finish_fields()
        continue
    found_fields.update({x.split(':')[0]: x.split(':')[1] for x in entry.split(' ')})
finish_fields()
print(count)

File: c++/day4b.cpp
#include <algorithm>
#include <cstring>
#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <sstream>
#include <string>
#include <vector>

using Passport = std::map<std::string, std::string>;
using Input = std::vector<Passport>;

Input readParseInput(std::string fileName) {
    std::ifstream file(fileName);
    Input input;
    Passport *mPtr;

    mPtr = &input.emplace_back();

    std::string line;
    while (std::getline(file, line)) {
        auto length = line.length();
        if (!length) {
            mPtr = &input.emplace_back();
            continue;
        }
        std::string::size_type parsePos = 0;
        while (true) {
            auto colonPos = line.find(':', parsePos);
            if (colonPos == std::string::npos)
                break;
            auto spacePos = line.find(' ', colonPos);
            if (spacePos == std::string::npos)
                spacePos = length;

            std::string fieldName = line.substr(parsePos, colonPos - parsePos);
            std::string value = line.substr(colonPos + 1, spacePos - (colonPos + 1));
            mPtr->insert({fieldName, value});

            if (spacePos == length)
                break;
            parsePos = spacePos + 1;
        }
    }
    file.close();

    return input;
}

using fieldValidityFunc = bool (*)(const std::string &);

bool validByr(const std::string &value) {
    if (value.length() != 4)
        return false;
    std::istringstream iss(value);
    int num;
    iss >> num;
    return (num >= 1920) && (num <= 2002);
}

bool validIyr(const std::string &value) {
    if (value.length() != 4)
        return false;
    std::istringstream iss(value);
    int num;
    iss >> num;
    return (num >= 2010) && (num <= 2020);
}

bool validEyr(const std::string &value) {
    if (value.length() != 4)
        return false;
    std::istringstream iss(value);
    int num;
    iss >> num;
    return (num >= 2020) && (num <= 2030);
}

bool validHgt(const std::string &value) {
    std::istringstream iss(value);
    int num;
    char units[3];
    iss >> num >> units;
    if (!strcmp(units, "cm"))
        return (num >= 150) && (num <= 193);
    else if (!strcmp(units, "in"))
        return (num >= 59) && (num <= 76);
    else
        return false;
}

bool validHcl(const std::string &value) {
    if (value.length() != 7)
        return false;
    if (value[0] != '#')
        return false;
    if (value.find_first_not_of("0123456789abcdef", 1) != std::string::npos)
        return false;
    return true;
}

const std::set<std::string> validEcls = {
    "amb",
    "blu",
    "brn",
    "gry",
    "grn",
    "hzl",
    "oth",
};

bool validEcl(const std::string &value) {
    return validEcls.find(value) != validEcls.cend();
}

bool validPid(const std::string &value) {
    if (value.length() != 9)
        return false;
    if (value.find_first_not_of("0123456789", 0) != std::string::npos)
        return false;
    return true;
}

const std::pair<std::string, fieldValidityFunc> fieldValidityChecks[] = {
    { "byr", validByr },
    { "iyr", validIyr },
    { "eyr", validEyr },
    { "hgt", validHgt },
    { "hcl", validHcl },
    { "ecl", validEcl },
    { "pid", validPid },
};

bool isValidPassport(const Passport &passport) {
    return std::all_of(
        std::begin(fieldValidityChecks),
        std::end(fieldValidityChecks),
        [passport] (const auto requiredField) {
            const auto &[fieldName, checkFunc] = requiredField;
            auto it = passport.find(fieldName);
            if (it == passport.cend())
                return false;
            const auto &fieldVal = it->second;
            return checkFunc(fieldVal);
        }
    );
}

int answer(const Input &input) {
    return std::ranges::count_if(input, [](const Passport &p) { return isValidPassport(p); });
}

int main(void) {
    std::cout << answer(readParseInput("../input/day4.txt")) << '\n';
    return 0;
}
